packages/orca-sdk/src/orca-lookup-fetcher/index.ts

```diff
@@ -0,0 +1,123 @@
+  private reverseAddressIndex: { [key: string]: Set<string> } = {};
+
+  constructor(readonly server: AxiosInstance, readonly connection: Connection) {}
+
+  async loadLookupTables(addresses: PublicKey[]): Promise<LookupTable[]> {
+    // Server request ~ 200ms
+
+    // Check whether any address is already indexed
+    const matchedIndex = addresses.map((addr) => this.reverseAddressIndex[addr.toString()] || null);
+
+    // Check to make sure either we've already missed a fetch on the address or that there is a match
+    let allMatched = true;
+    for (let i = 0; i < addresses.length; i++) {
+      const matched = matchedIndex[i];
+      if (matched == null && !this.localCacheMiss.has(addresses[i].toString())) {
+        allMatched = false;
+        break;
+      }
+    }
+
+    // If all are matched or previously missed, use the cache
+    if (allMatched) {
+      const allLuts = [];
+      for (let i = 0; i < matchedIndex.length; i++) {
+        const rai = matchedIndex[i];
+        if (rai) {
+          allLuts.push(...Array.from(rai!));
+        }
+      }
+      const lutSet = new Set(allLuts);
+      return Array.from(lutSet).map((lut) => this.localLutCache[lut]);
+    }
+
+    // Otherwise, lookup all addresses from the server, and update the cache
+    const response = await this.server.request({
+      url: "/lookup-table",
+      method: "get",
+      params: {
+        addresses: addresses.map((addr) => addr.toBase58()).join(","),
+      },
+    });
+
+    const lookupTables: LookupTable[] | undefined = response?.data?.lookupTables.map(
+      (lut: OrcaLookupTableModel) => {
+        return {
+          address: lut.address,
+          containedAddresses: lut.addresses,
+        };
+      }
+    );
+    if (lookupTables) {
+      this.updateCache(addresses, lookupTables);
+      return lookupTables;
+    }
+    return [];
+  }
+
+  async getLookupTableAccountsForAddresses(
+    addresses: PublicKey[]
+  ): Promise<AddressLookupTableAccount[]> {
+    const luts = await this.loadLookupTables(addresses);
+    return (
+      await Promise.all(
+        luts.map(async (lut) => {
+          if (this.resolvedAltCache[lut.address]) {
+            return Promise.resolve(this.resolvedAltCache[lut.address]);
+          } else {
+            const alt = (await this.connection.getAddressLookupTable(new PublicKey(lut.address)))
+              .value;
+            if (alt != null) {
+              this.resolvedAltCache[lut.address] = alt;
+            }
+            return alt;
+          }
+        })
+      )
+    ).filter((alt) => alt != null) as AddressLookupTableAccount[];
+  }
+
+  private updateCache(addresses: PublicKey[], lookupTables: LookupTable[]) {
+    // If we find lookup tables, update local caches
+    for (const lut of lookupTables) {
+      const { address, containedAddresses } = lut;
+      this.localLutCache[address] = lut;
+
+      // Create a reverse index from contained address => lookupTable
```

Comment:
i'll leave a comment here where the resolvedAltCache needs a TTL.
For our current use-cases on UI, the cache is short-lived so this isn't much of an issue.
For future use-cases, this can be an issue but the error will not fail explicitly. If the cached account is out of date, the transactions will likely not receive the benefits of reduced transaction-size and therefore fail w/ a tx too large error.

Reply:
Is there 1 ALT assigned to each pool now?
In that case, it is fine.

I was a little concerned about the following case.

```
ALT1: containedAddresses: [P, A, B, C, D]
ALT2: containedAddresses: [P, C, D, E, F]
ALT3: containedAddresses: [P, E, F, G, H]
ALT4: containedAddresses: [P, G, H, I, J]

1. loadLookupTable([A, B, C]) --> ALT1 and ALT2 will be loaded and P,A,B,C,D,E,F will be considered as matched.
                              --> Strictly speaking, only A, B, and C are fully cached
2. loadLookupTable([E, F]) --> all matched (ALT3 will NOT be loaded)
```

Reply:
Actually reviewing the ALT code, TTL isn't necessary because ALT's ownership of tick-array indices do not change. The sequence of tick-arrays of a pool is represented by a range of ALTs. The tick-arrays represented in each ALT is pre-determined by it's index and will not change. However, most of the time, the usable range of a pool other than TS1 pools will only stay within 1 ALT.

I may not be catching on to the case you are referring to for loading adjacent tick-arrays. Could you elaborate more?

The quote explicitly defines the tick-arrays that it used to complete the quote. If the router.findAllRoutes call only iterated to [E, F] for this pool, then returning ALT2 will satisfy 
`loadLookupTable([E, F])`.

If you are referring to the case where the quote iterated from [E, F, G] & the cache has [E], [F] matched to ALT2, then `loadLookupTable([E, F, G]) would invoke a server search on [E,F,G] since G is missing. It'll return ALT2 & ALT3 and update the cache accordingly.

Reply:
Thanks for the detailed explanation.
Now that I understand the entries contained in the ALT, my concerns are gone. 👍